dubbo-remoting/dubbo-remoting-zookeeper/src/main/java/org/apache/dubbo/remoting/zookeeper/support/AbstractZookeeperTransporter.java

```diff
@@ -0,0 +1,130 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.remoting.zookeeper.support;
+
+import org.apache.dubbo.common.Constants;
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.logger.Logger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperClient;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperTransporter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * 2019/1/10
+ */
+public abstract class AbstractZookeeperTransporter implements ZookeeperTransporter {
+    static final Logger logger = LoggerFactory.getLogger(ZookeeperTransporter.class);
+    Map<String, ZookeeperClientData> zookeeperClientMap = new HashMap<String, ZookeeperClientData>();
+
+    /**
+     * share connnect for registry, metadata, etc..
+     *
+     * @param url
+     * @return
+     */
+    public ZookeeperClient connect(URL url) {
+        ZookeeperClientData clientData;
+        List<String> addressList = getURLBackupAddress(url);
+        // The field define the zookeeper server , including protocol, host, port, username, password
+        if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+            logger.info("ZookeeperTransporter.connnect result from map:" + clientData);
+            return clientData.getZookeeperClient();
+        }
+        ZookeeperClient zookeeperClient = null;
+        synchronized (zookeeperClientMap) {
+            if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+                logger.info("ZookeeperTransporter.connnect result from map2:" + clientData);
+                return clientData.getZookeeperClient();
+            }
+
+            Set<URL> originalURLs = new HashSet<>(2);
+            originalURLs.add(url);
+            zookeeperClient = createZookeeperClient(createServerURL(url), originalURLs);
+
+            ZookeeperClientData zookeeperClientData = new ZookeeperClientData(zookeeperClient, originalURLs);
+            logger.info("ZookeeperTransporter.connnect result from new connection. " + clientData);
+
+            writeToClientMap(addressList, zookeeperClientData);
+        }
+        return zookeeperClient;
+    }
+
+    protected abstract ZookeeperClient createZookeeperClient(URL url, Set<URL> originalURLs);
+
+    ZookeeperClientData fetchAndUpdateZookeeperClientCache(URL url, List<String> addressList) {
```

Comment:
method access specifier plz.

Reply:
it is can be private method.  But I want this method can be unit test.